src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/MultiStringMatcher.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+
+namespace System.Text.RegularExpressions
+{
+    /// <summary>
+    /// Efficiently finds a set of strings in text.
+    /// Currently uses the Aho-Corasick algorithm.
+    /// </summary>
+    internal sealed class MultiStringMatcher
+    {
+        private readonly TrieNodeWithLinks[] _trie;
+
+        internal ReadOnlySpan<TrieNodeWithLinks> Trie => _trie;
+
+        /// <summary>
+        /// Creates a <see cref="MultiStringMatcher"/> from a trie, and
+        /// calculates its nodes' suffix and dictionary links.
+        /// </summary>
+        internal MultiStringMatcher(List<TrieNode> trie)
+        {
+            _trie = BuildTrieLinks(trie);
+        }
+
+        private static TrieNodeWithLinks[] BuildTrieLinks(List<TrieNode> trie)
+        {
+            TrieNodeWithLinks[] result = new TrieNodeWithLinks[trie.Count];
+            Queue<int> vertexQueue = new Queue<int>();
+            vertexQueue.Enqueue(TrieNode.Root);
+#if REGEXGENERATOR
+            while (vertexQueue.Count > 0)
+            {
+                int currentVertex = vertexQueue.Dequeue();
+#else
+            while (vertexQueue.TryDequeue(out int currentVertex))
+            {
+#endif
+                TrieNode node = trie[currentVertex];
+                int suffixLink;
+                int matchLength;
+
+                if (node.IsRoot)
+                {
+                    suffixLink = TrieNode.Root;
+                    matchLength = -1;
+                    goto End;
+                }
+
+                // one character substrings
+                if (node.Parent == TrieNode.Root)
+                {
+                    suffixLink = TrieNode.Root;
+                    matchLength = node.IsMatch ? node.Depth : result[suffixLink].MatchLength;
+                    goto End;
+                }
+
```

Comment:
This doesn't need to use gotos... it can just be:
```C#
if (node.IsRoot)
{
    suffixLink = TrieNode.Root;
    matchLength = -1
}
else if (node.Parent == TrieNode.Root)
{
    suffixLink = TrieNode.Root;
    matchLength = node.IsMatch ? node.Depth : result[suffixLink].MatchLength;
}
else
{
    // To calculate the suffix link for the current vertex, we need the suffix
    ...
}

result[currentVertex] = new TrieNodeWithLinks()
...
```